buildkite/resource_pipeline_test.go

```diff
@@ -98,6 +98,127 @@ func TestAccPipeline_add_remove_withcluster(t *testing.T) {
+				),
+			},
+		},
+	})
+}
+
+// TestAccPipeline_add_remove_withcluster_old_version tests that the provider doesn't get stuffed up when coming from an
+// older version to this new version. the first step runs against the published v0.21.1 version and the second step runs
+// against the current provider. the configuration is the same between each one so there should be no state/plan change
+func TestAccPipeline_add_remove_withcluster_old_version(t *testing.T) {
+	var resourcePipeline PipelineNode
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		CheckDestroy: testAccCheckPipelineResourceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccPipelineConfigBasicWithCluster("foo"),
+				ExternalProviders: map[string]resource.ExternalProvider{
+					"buildkite": {
+						Source:            "registry.terraform.io/buildkite/buildkite",
+						VersionConstraint: "0.21.1",
+					},
+				},
+				Check: resource.ComposeAggregateTestCheckFunc(
+					// Confirm the pipeline exists in the buildkite API
+					testAccCheckPipelineExists("buildkite_pipeline.foobar", &resourcePipeline),
+					// Confirm the pipeline has the correct values in Buildkite's system
+					testAccCheckPipelineRemoteValues(&resourcePipeline, "Test Pipeline foo"),
+					// Confirm the pipeline has the correct values in terraform state
+					resource.TestCheckResourceAttr("buildkite_pipeline.foobar", "name", "Test Pipeline foo"),
+					resource.TestCheckResourceAttr("buildkite_pipeline.foobar", "cluster_id", "Q2x1c3Rlci0tLTFkNmIxOTg5LTJmYjctNDRlMC04MWYyLTAxYjIxNzQ4MTVkMg=="),
+					resource.TestCheckResourceAttr("buildkite_pipeline.foobar", "allow_rebuilds", "true"),
+					// check provider settings are present
+					resource.TestCheckResourceAttr("buildkite_pipeline.foobar", "provider_settings.0.trigger_mode", "code"),
+				),
+			},
+			{
+				Config:                   testAccPipelineConfigBasicWithCluster("foo"),
+				ProtoV6ProviderFactories: protoV6ProviderFactories(),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					// Confirm the pipeline exists in the buildkite API
+					testAccCheckPipelineExists("buildkite_pipeline.foobar", &resourcePipeline),
+					// Confirm the pipeline has the correct values in Buildkite's system
+					testAccCheckPipelineRemoteValues(&resourcePipeline, "Test Pipeline foo"),
+					// Confirm the pipeline has the correct values in terraform state
+					resource.TestCheckResourceAttr("buildkite_pipeline.foobar", "name", "Test Pipeline foo"),
+					resource.TestCheckResourceAttr("buildkite_pipeline.foobar", "cluster_id", "Q2x1c3Rlci0tLTFkNmIxOTg5LTJmYjctNDRlMC04MWYyLTAxYjIxNzQ4MTVkMg=="),
+					resource.TestCheckResourceAttr("buildkite_pipeline.foobar", "allow_rebuilds", "true"),
+					// check provider settings are present
+					resource.TestCheckResourceAttr("buildkite_pipeline.foobar", "provider_settings.0.trigger_mode", "code"),
```

Comment:
This asserts that provider settings are present in state. There is a separate test that ensures its present in state when its present in the resource attributes. But this test doesn't specify the values so it doesn't put it into state. This test is currently failing because of that.

Previous versions of the provider put this into state regardless of whether the user provided a `provider_settings` block. This is somewhat an artifact/bug of the old SDKv2 that this was supposed to be required and makes sense for it to be required I think. The API will set default values on the pipeline settings when it creates a pipeline, but those aren't properly reflected in state and is the source of a few open issues we have currently.

I think long term, provider settings should be made required and not a list but a single block attribute (or nested attribute in protocol v6). I'm unsure whether to do that as part of this or at all. Or whether to achieve parity with the SDK version of this resource.

Reply:
I think the way to approach this is to put as much effort into parity here with just the migration from SDKv2 to framework. But I think its worth looking at the state schema afterwards to clean it up, prior to a v1 release of the provider. Hopefully the new state migration functionality in framework will make that easier.

Reply:
> I think long term, provider settings should be made required

I don't agree with this, or, I think it's maybe the easiest approach but not the most user focused; making them required is fine in a sense that we're then able to be hands off, in a way, but folks who won't be using `code`, or source control for the trigger will find this convoluted. I think maybe we can find an alternative solution, though that may require additional engineering in the GraphQL API